### ISSUE TITLE
Issue with Calix E7 Interface Polling/Graphing

### ISSUE THREAD

**Issue by thecityofguanyu** [librenms/librenms#3222]
Daily update enabled, current verison: <code>2e802d8</code>

As of now I have three Calix E7 systems added as nodes. None of the are accurately graphing the traffic that is passing across the interfaces.

You'll see below that the realtime graph is displaying very incorrect thoughput values when compared to the actual admitted traffic on a GPON port:
![realtime]
(http://i.imgur.com/LkoaDrA.png)

The highlighted row in the image below shows the values in the librenms database for ifInOctets/ifOutOctets on one particular GPON port:
![port-db]
(http://i.imgur.com/0QXc402.png)

Yet the graph for that port looks as follows:
![port-graph]
(http://i.imgur.com/AdvlsTm.png)

Maybe I'm mistaken here, but shouldn't it be taking the ifHCInOctets & ifHCOutOctets? As opposed to just ifInOctets/ifOutOctets. I'm probably wrong there, as I can't find high capacity counter info anywhere in the database for our other equipment (Brocade, Juniper, Cisco) and they all graph correctly.

If it helps, we were able to get this to successfully graph on our old Cacti deployment with the values configured below; however, we're trying to move away from Cacti over to LibreNMS.
![cacti1]
(http://i.imgur.com/bIPLh2y.png)

Any ideas here?

@murrant tagging you because we discussed it briefly on IRC and I think you added the E7-CALIX MIB.

**EDIT**:

Regarding 
>Maybe I'm mistaken here, but shouldn't it be taking the ifHCInOctets & ifHCOutOctets?

Seems to be a pretty significant difference between
```
snmpwalk -v2c -c public -M +mibs -m +all 10.1.10.27 ifHCOutOctets.30101
IF-MIB::ifHCOutOctets.30101 = Counter64: 10142998413825
```
And
```
snmpwalk -v2c -c public -M +mibs -m +all 10.1.10.27 ifOutOctets.30101
IF-MIB::ifOutOctets.30101 = Counter32: 2581954566
```
The value of ifHCOutOctets appears to match the DB table image in the original post very closely while ifOutOctets does not, so I would wager that it is correctly polling 64-bit counters and adding that data to the database.

Why it is not graphing properly is still beyond me.

**Comment by thecityofguanyu** [comment]
Fairly significant difference between:
```
snmpwalk -v2c -c public -M +mibs -m +all 10.1.10.27 ifHCOutOctets.30101
IF-MIB::ifHCOutOctets.30101 = Counter64: 10142998413825
```
And

``` 
snmpwalk -v2c -c public -M +mibs -m +all 10.1.10.27 ifOutOctets.30101
IF-MIB::ifOutOctets.30101 = Counter32: 2581954566
```

The value of ifHCOutOctets appears to match the DB table image in the original post very closely while ifOutOctets does not, so I would wager that it is correctly polling 64-bit counters and adding that data to the database.

**Comment by murrant** [comment]
@thecityofguanyu Can you try using SNMPv3 for one of your E7s and see if that makes a difference.

**Comment by thecityofguanyu** [comment]
@murrant It definitely still looks off.

Admitted rate:
```
Bandwidth for GPON port "1/1/1": 

Aggregate for PON:
Upstream                                Downstream
--------                                ----------
Guaranteed max rate       : 1244 Mb/s   Policed max rate        : 2488 Mb/s
Guaranteed admitted rate  : 0 Mb/s      Policed admitted rate   : 557 Mb/s
Guaranteed available rate : 1243 Mb/s   Policed available rate  : 1930 Mb/s
Guaranteed % of PON       : 0%          Policed % of PON        : 22%
Guaranteed % of max       : 0%          Policed % of max        : 22%
Best effort max rate      : 1243 Mb/s
Best effort admitted rate : 163 Mb/s
Best effort available rate: 1080 Mb/s
Best effort % of PON      : 13%
Best effort % of max      : 13%
```
Versus what is graphed:
![snmpv3]
(http://i.imgur.com/nQAM6sX.png)

>rrdtool graph /tmp/l1mXrRnVd8HMdddi.png --alt-autoscale-max --rigid -E --start 1457881800 --end 1457968200 --width 860.4 --height 483.5 -c BACK#EEEEEE00 -c SHADEA#EEEEEE00 -c SHADEB#EEEEEE00 -c FONT#000000 -c CANVAS#FFFFFF00 -c GRID#a5a5a5 -c MGRID#FF9999 -c FRAME#5e5e5e -c ARROW#5e5e5e -R normal --font LEGEND:8:DejaVuSansMono --font AXIS:7:DejaVuSansMono --font-render-mode normal DEF:outoctets=/opt/librenms/rrd/10.1.10.27/port-id29201.rrd:OUTOCTETS:AVERAGE DEF:inoctets=/opt/librenms/rrd/10.1.10.27/port-id29201.rrd:INOCTETS:AVERAGE DEF:outoctets_max=/opt/librenms/rrd/10.1.10.27/port-id29201.rrd:OUTOCTETS:MAX DEF:inoctets_max=/opt/librenms/rrd/10.1.10.27/port-id29201.rrd:INOCTETS:MAX CDEF:octets=inoctets,outoctets,+ CDEF:doutoctets=outoctets,-1,* CDEF:outbits=outoctets,8,* CDEF:outbits_max=outoctets_max,8,* CDEF:doutoctets_max=outoctets_max,-1,* CDEF:doutbits=doutoctets,8,* CDEF:doutbits_max=doutoctets_max,8,* CDEF:inbits=inoctets,8,* CDEF:inbits_max=inoctets_max,8,* VDEF:totin=inoctets,TOTAL VDEF:totout=outoctets,TOTAL VDEF:tot=octets,TOTAL VDEF:95thin=inbits,95,PERCENT VDEF:95thout=outbits,95,PERCENT CDEF:d95thoutn=doutbits,-1,* VDEF:d95thoutn95=d95thoutn,95,PERCENT CDEF:d95thoutn95n=doutbits,doutbits,-,d95thoutn95,-1,*,+ VDEF:d95thout=d95thoutn95n,FIRST COMMENT:'bps Now Ave Max 95th %\n' AREA:inbits_max#D7FFC7: AREA:inbits#90B040: LINE:inbits#608720:'In ' GPRINT:inbits:LAST:%6.2lf%s GPRINT:inbits:AVERAGE:%6.2lf%s GPRINT:inbits_max:MAX:%6.2lf%s GPRINT:95thin:%6.2lf%s\n AREA:doutbits_max#E0E0FF: AREA:doutbits#8080C0: LINE:doutbits#606090:'Out' GPRINT:outbits:LAST:%6.2lf%s GPRINT:outbits:AVERAGE:%6.2lf%s GPRINT:outbits_max:MAX:%6.2lf%s GPRINT:95thout:%6.2lf%s\n GPRINT:tot:'Total %6.2lf%sB' GPRINT:totin:'(In %6.2lf%sB' GPRINT:totout:'Out %6.2lf%sB)\l' LINE1:95thin#aa0000 LINE1:d95thout#aa0000

The database matches what is graphed at least in some regard, as 563743 (in the ifOutOctets_rate field) matches 4.5 Mb/s.
![database]
(http://i.imgur.com/83eyh94.png)

The question is, how is it getting 4.5 Mb/s instead of ~550 Mb/s as it should?

**Comment by thecityofguanyu** [comment]
@murrant 

No difference when polling with SNMPv3. This does seem to impact all 3 E7s.

Strange (impossible 17Gbps) results such as the rates seen here:
![17gbps]
(http://i.imgur.com/y2jEyni.png)

And this graph is all over the place:
![graph]
(http://i.imgur.com/Q5I5K5C.png)

Sometimes the graphs spike to hunders of Gb/s values to even Tb/s values for seemingly no rhyme or reason.

I have deleted the .rrd files for the hosts and let them re-generate, but the issues just re-occur.

The actual values don't necessarily seem concerning when I snmpwalk and seem fine when I check them in the database, so I'm not sure what is happening to these graphs.

**Comment by murrant** [comment]
@thecityofguanyu is your poller completing within 5minutes?

**Comment by thecityofguanyu** [comment]
@murrant I've extended the time, as well as added a few more E7s for it to poll against. I'll give it some time and check back. Thanks for the help.

**Comment by thecityofguanyu** [comment]
I've determined this to be an issue with E7 systems themselves. It seems to impact modular systems with a very large number of ports needed to poll (1000+). High CPU usage alarms come in on said devices as well, as a result of the polling. Calix has identified some bug related to snmpwalk-ing E7s on the code version we're running. I will attempt to re-enable ports polling on such devices once we are upgraded to the recommended release version.

**Comment by murrant** [comment]
Out of curiosity, what version did they reference?  2.4?

**Comment by thecityofguanyu** [comment]
@murrant Issue reported in "earlier" 2.3 release, resolved in 2.3.50.

>System
? Resolved: When performing an SNMPwalk on an E7-2 GPON system, a low-sw-alarm may occur.
(EX-18139, EX-17546, 17513)

**Comment by murrant** [comment]
Ah, we are running >=2.3.50 that must be why I didn't see it.
